man/dbload.1
.TH DBLOAD 1 " 4 April 1996"
.FB
.SH NAME
dbload, dbrload \- load comma separated values into a \fIFirstBase\fP database
.SH SYNOPSIS
.B dbload
[\fB-b\fP]\ [\fB-y\fP]\ [\fB-a\fP] [\fB-l\fP]
[\fB-d\fP\ \fIdatabase\fP] [\fB-c\fP\ \fIseparator\fP]
[\fIloadfile\fP]
.br
.B dbrload
[\fB-b\fP] [\fB-y\fP]
[\fB-d\fP \fIdatabase\fP]
[\fB-i\fP \fIindex\fP]
[\fB-c\fP \fIseparator\fP]
[\fIloadfile\fP]
.PD
.SH DESCRIPTION
.I Dbload
reads comma separated values from 
.I loadfile
and builds a \fIFirstBase\fP database
using the values for fields. Each line of the input is assumed to be one
record. A valid \fIFirstBase\fP database dictionary is required.
.PP
\fIDbload\fP(1) can also be used with an existing database. Using the
\fB-a\fP option will cause \fIdbload\fP(1) to append records from the
loadfile to the database. The database map is also updated, so use of
\fIdbclean\fP(1) is not required. To enforce field length checking,
use the \fB-l\fP option.
.PP
.I Dbrload
is similar to dbload, except that database records are replaced (updated)
in an already existing database. To use \fIdbrload\fP(1), at least the
database must exist already. For each line of the loadfile,
\fIdbrload\fP(1) will
determine if the record already exists by consulting the given index,
which must be a FirstBase autoindex.
.PP
If the record exists, all non blank fields from the loadfile
are written to the existing database record.
To force the clearing of a field, use \fB$ERASE\fP as the field value.
If the record does not exist, or if no index is given for \fIdbrload\fP(1)
to search, a new record is appended to the end of the database.
Again: without an index, \fIdbrload\fP(1) will append
all records from the loadfile as if they were new.
.PP
To work properly on repeated duplicate records in the load file,
\fIdbrload\fP(1) must update the
given index as it adds new records to the database.
The best use of this feature is with \fIFirstBase\fP Btree indexes
which retain their order during record insertions.
.PP
Note that \fIdbrload\fP(1) will work on complex indexes \(em and the fields
that make up this index do not need to be in any specific order.
However, only one non-auto index will be updated by \fIdbrload\fP(1).
.PP
\fIDbrload\fP(1) also accepts the use of formulas for numeric fields.
These formulas have
the same basic format as those for \fIFirstBase\fP formula (virtual) fields
defined via \fIdbdbas\fP(1).
However, since the string passed in via dbrload could merely look like a
formula, a special mechanism is used to tell \fIdbrload\fP(1)
that the string is
to be parsed as a formula \(em all dbrload formulas must begin with "$F:".
.PP
.I Formulas 
use numbers to represent fields (eg. 16 for field 16)
and a prepended 'C' or 'c' for constants (eg. C4.5 for 4.5).
Addition 
.I (+), 
subtraction 
.I (-), 
multiplication 
.I (*), 
and division
.I (/)
are allowed as operators. 
.PP
.I Note:
Only dollar, float, numeric and positive numeric
fields can be used in formula
calculations. This means references to other formulas will not work.
Also, all formulas read from left to right, and there is no precedence
among the operators.
.PP
The precision of a formula is specified by appending a colon 
.I (:)
to the formula followed by a number.
Default precision is two.
.PP
For example, if the value of the third comma separated value of the rload
file is \fB$F:3+C22:0\fP,
then the contents of field three will be increased by
22 \(em with the result having 0 places after the decimal.
.PP
The \fB-b\fP
flag will run the \fIFirstBase\fP tool in batch mode, suitable for shell
scripts.
This mode produces no output on the terminal, thus can also be used
when running the process in the background.
The \fB-y\fP flag will run the tool at the terminal without pausing to ask
the user if existing files can be destroyed and without displaying
a 'picture' of the requested task.
.PP
If 
.I loadfile 
is not specified, the user is prompted for a
file name. The default is \fI*.emit\fP, where * is the database file name.
.I Note: 
normally, 
\fIdbload\fP(1) 
will 
.I not
add the extension \fI.emit\fP to the
.I loadfile.
.PP
If there are more fields defined in the dictionary than there are
values in any one input line, 
the excess fields are null padded. Excess values
not defined in the dictionary are ignored.
.PP
Although it is not necessary to put alphanumeric data in quotes, this will
not hinder dbload(1). Quotes must be used if the data includes
embedded commas. The backslash (\\)
is a metacharacter used to escape characters.
To embed a quotation mark, use '\\"', 
and to embed a backslash, '\\\\'.
.PP
.I Dbload(1)
provides a convenient method of uploading from one 
database management system into the
.I FirstBase
system. The 'comma separated field'
format is compatible with many other database
management packages.
It is also possible to emit values out of a \fIFirstBase\fP database file. See
\fIdbemit\fP(1).
.PP
The \fB-c\fP argument can be used to change the field separator to another
character. For example, \fB-c \\|\fP could be used to change the separator to
the vertical bar character ('|'). (Note the escaping of the pipeline
character due to the shell.)
.PP
Note: Unless \fB-a\fP was used, \fIdbclean\fP(1)
is automatically run at the completion of \fIdbload\fP(1) to
create a map file for the new database.
.SH FILES
.PD 0
.TP 10
dbase
default name of database.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.map
record map of \fIFirstBase\fP database.
.TP 10
*.ddict
\fIFirstBase\fP database dictionary file.
.TP 10
*.emit
\fIFirstBase\fP database emit file (dbemit(1)), defaulted to for dbload(1).
.PD
.SH SEE ALSO
dbcat(1), dbdbas(1), dbemit(1), dbjoin(1), firstbase(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
\fIDbload\fP(1)
will ask permission before overwriting an existing \fIFirstBase\fP database.
.sp 1
Note: \fIdbrload\fP(1) tool will attempt to gain write access to a record
before operating on that record. If access is currently locked, then the
internal locking routine will respond according to the value of
the \fIFirstBase\fP
environment variable LOCKMESSAGE. See \fIsetup\fP(5).
.sp
Each of these tools normally will ignore interrupt signals. However, if
the firstbase variable INTERRUPT is ON, these tools will terminate on
an interrupt signal.
However, if either of these database load tools is interrupted,
the database and index objects being updated should be deleted using
\fIrmdb\fP(8) and \fIrmidx\fP(8) before the database load is attempted again.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbtools/dbload.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbload.c,v 9.1 2001/02/16 19:47:51 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Dbload_sid[] = "@(#) $Id: dbload.c,v 9.1 2001/02/16 19:47:51 john Exp $";
#endif

/* 
 *  dbload.c - load a file of fb_field values in comma seperated format 
 *     into a cdb fb_database. needs the dictionary to load describe
 *     how to load the values.
 */

#include <fb.h>
#include <fb_vars.h>
#include <emit.h>

#if !FB_PROTOTYPES
static int load();
static int aload();
static int simplefield();
static finit();
static void usage();
#else /* FB_PROTOTYPES */
extern main(int, char **);
static int load(void);
static int aload(void);
static int simplefield(void);
static finit(int argc, char **argv);
static void usage(void);
#endif /* FB_PROTOTYPES */

extern char *cdb_DBCLEAN;
extern short int cdb_secure;
extern char *cdb_S_EOREC;
extern char *cdb_S_FILLC;
extern char *cdb_S_FILLE;
extern short int cdb_dbase_byteorder;
extern short int cdb_cpu_byteorder;

static char *FMT1 = "%04d%03d%s";	/* secure bits */

static char lastc = NULL;		/* last char seen - pushback kinda */
static char separator = ',';		/* standard separator */
static append = 0;			/* -a flag for fast append */
static long rlen = 0L, size = 0L;
static lengthcheck = 0;

extern short int cdb_interrupt;

/*
 *  dbload - main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
   
      {
         char str[FB_MAXLINE], com[FB_MAXLINE];
	 int j, st, ddflag = 0;

         (void) Dbload_sid;

         fb_getargs(argc, argv, FB_ALLOCDB);
	 if (cdb_interrupt)
            fb_allow_int();
	 if (fb_testargs(argc, argv, "-dd") > 0)
	    ddflag = 1;
         if (fb_getd_dict(cdb_db) == FB_ERROR)
            fb_xerror(FB_BAD_DICT, cdb_db->ddict, NIL);
	 if ((j = fb_testargs(argc, argv, "-a")) > 0){
	    append = 1;
	    fb_opendb(cdb_db, READWRITE, FB_NOINDEX, 0);
	    }
         lengthcheck = 1;
         fb_getauto(cdb_db, READWRITE);
         fb_scrhdr(cdb_db, "Open Files"); fb_infoline();
	 if ((j = fb_testargs(argc, argv, "-c")) > 0){
	    if (++j >= argc)
	       usage();
	    if (argv[j][1] != NULL)
	       usage();
	    separator = argv[j][0];
	    }
	 finit(argc, argv);
	 if (!cdb_batchmode && !cdb_yesmode){
	    if (!append)
	       sprintf(str, "Ok to Load into FirstBase Database %s (y/n) ?", 
		  cdb_db->dbase);
	    else
	       sprintf(str, "Ok to Append to FirstBase Database %s (y/n) ?", 
		  cdb_db->dbase);
	    if (fb_mustbe('y', str, cdb_t_lines, 1) == FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
	 if (append)
	    st = aload();
	 else
	    st = load();
         if (st == FB_ERROR)
	    fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	 if (append){
	    fb_closedb(cdb_db);
	    }
	 else {
	    fb_rootname(str, cdb_db->dbase);
	    if (!cdb_batchmode){
	       fb_move(4,1), fb_clrtobot();
	       fb_scrstat("Cleaning");
	       sprintf(com, "Cleaning %s", str);
	       fb_fmessage(com);
	       }
	    sprintf(com, "%s -b %s", cdb_DBCLEAN, str);
            if (ddflag){
               fb_rootname(str, cdb_db->ddict);
               sprintf(com, "%s -dd %s", com, str);
               }
	    fb_system(com, FB_WITHROOT);
	    }
	 fb_r_end();
	 fb_ender();
      }

/* 
 *  load - load all lines from a load file (fd) into a FirstBase database 
 */
 
   static int load()
      
      {
         long n;
	 int i, eof, st, nc, line_no = 1;
         char perm[FB_MAXNAME], tbuf[FB_MAXLINE];

	 n = 0L;
         if (!cdb_batchmode){
	    FB_XWAIT();
	    fb_gcounter(n);
	    }
	 close(creat(cdb_db->dbase, 0666));		/* create the dbase */
         cdb_db->fd = fb_mustopen(cdb_db->dbase, 1);
	 fb_putseq(cdb_db->fd);
	 fb_puthead(cdb_db->fd, 0L, 0L);
         if (cdb_secure){
            if (fb_putmode(cdb_db->fd, fb_getuid(), fb_getgid(),
                  "666") == FB_ERROR)
               fb_xerror(FB_IO_ERROR, cdb_db->dbase, NIL);
	    sprintf(perm, FMT1, fb_getuid(), fb_getgid(), "666");
            }
	 fb_w_init(1, cdb_db->fd, -1);
	 for (i = 0, eof = 0 ; ; ){
	    for (i = 0; i < cdb_db->nfields; i++)
	       if (cdb_kp[i]->type != FB_FORMULA &&
                     cdb_kp[i]->dflink == NULL && cdb_kp[i]->type != FB_LINK){
	          if (lastc != FB_NEWLINE){
		     if (simplefield() == EOF){	/* returns in cdb_bfld */
			eof = 1;
			break;
			}
                     if (cdb_kp[i]->type != FB_BINARY){
                        if (lengthcheck && strlen(cdb_bfld) > cdb_kp[i]->size){
	                   sprintf(tbuf, "Truncating Field %d of line %d",
		              i + 1, line_no);
                           cdb_bfld[cdb_kp[i]->size] = NULL;
	                   fb_serror(FB_MESSAGE, tbuf, NIL);
                           }
		        fb_nextwrite(0, cdb_bfld);
                        }
		     }
                  if (cdb_kp[i]->type != FB_BINARY)
		     fb_w_write(0, "\000");		/* EOS = e-o-field */
                  else{					/* pad BINARY field */
                     for (nc = cdb_kp[i]->size; nc > 0; nc--)
                        fb_w_write(0, "\000");
                     }
		  }
	    if (lastc != FB_NEWLINE){
	       for (;;){
	          if ((st = fb_nextread(&lastc)) == 0)
		     break;
		  if (lastc == FB_NEWLINE)
		     break;
		  }
	       if (st == 0){
		  eof = 1;
		  break;
		  }
	       }
	    if (eof == 1)
	       break;
	    fb_w_write(0, " ");
            if (cdb_secure)
 	       fb_nextwrite(0, perm);		      	/* permissions */
	    fb_w_write(0, "\000");			/* EOS = e-o-field */
	    fb_w_write(0, cdb_S_EOREC);			/* end of record */
	    lastc = NULL;
	    n++;
            line_no++;
	    if (!cdb_batchmode)
	       fb_gcounter(n);
	    }
	 fb_wflush(1);
	 if (fb_puthead(cdb_db->fd, n, 0L) == FB_ERROR)
	    fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_HEADER], cdb_db->dbase);
	 fb_w_end(1);
         return(FB_AOK);
      }

/* 
 *  aload - append load all lines from a load file (fd) into database 
 */
 
   static int aload()
      
      {
         long n, avail, rpos, freep, w_avail, w_rlen;
	 int i, eof, st, nc, line_no = 1, length;
         char perm[FB_MAXNAME], tbuf[FB_MAXLINE];

         if (cdb_secure)
	    sprintf(perm, FMT1, fb_getuid(), fb_getgid(), "666");

	 if (!cdb_batchmode)
	    FB_XWAIT();

	 /* get header information from fb_database map */
	 if (fb_getmap(cdb_db->mfd, 0L, &avail, &freep, &rpos, &size) ==
               FB_ERROR){
	    fb_serror(FB_IO_ERROR, SYSMSG[S_BAD_MAP], NIL);
	    return(FB_ERROR);
	    }

         /* seek fb_database fd to avail, the end of the fb_database */
	 if (lseek(cdb_db->fd, avail, 0) < 0L){
	    fb_serror(FB_SEEK_ERROR, SYSMSG[S_BAD_DATA], NIL);
	    return(FB_ERROR);
	    }

         /* set the fd for the map to the end of the map */
	 n = cdb_db->reccnt;
	 if (lseek(cdb_db->mfd, (n + 1L) * (long) FB_MAPREC_SIZE, 0) < 0L){
	    fb_serror(FB_SEEK_ERROR, SYSMSG[S_BAD_MAP], NIL);
	    return(FB_ERROR);
	    }

         fb_w_init(2, cdb_db->fd, cdb_db->mfd);
	 for (eof = 0 ; ; ){
	    for (i = 0, size = 0L; i < cdb_db->nfields; i++)
	       if (cdb_kp[i]->type != FB_FORMULA &&
                     cdb_kp[i]->dflink == NULL && cdb_kp[i]->type != FB_LINK){
	          if (lastc != FB_NEWLINE){
		     if (simplefield() == EOF){	/* returns in cdb_bfld */
			eof = 1;
			break;
			}
                     if (cdb_kp[i]->type != FB_BINARY){
                        if (lengthcheck){
                           length =  strlen(cdb_bfld);
                           if (length > cdb_kp[i]->size){
	                      sprintf(tbuf, "Truncating Field %d of line %d",
		                 i + 1, line_no);
                              cdb_bfld[cdb_kp[i]->size] = NULL;
                              size -= (length - cdb_kp[i]->size);
	                      fb_serror(FB_MESSAGE, tbuf, NIL);
                              }
                           }
		        fb_nextwrite(0, cdb_bfld);
                        }
		     }
                  if (cdb_kp[i]->type != FB_BINARY){
		     fb_w_write(0, "\000");	/* EOS = e-o-fb_field */
		     size++;
                     }
                  else{				/* pad FB_BINARY fb_field */
                     for (nc = cdb_kp[i]->size; nc > 0; nc--)
                        fb_w_write(0, "\000");
                     size += cdb_kp[i]->size;
                     }
		  }
	    if (lastc != FB_NEWLINE){
	       for (;;){
	          if ((st = fb_nextread(&lastc)) == 0)
		     break;
		  if (lastc == FB_NEWLINE)
		     break;
		  }
	       if (st == 0){
		  eof = 1;
		  break;
		  }
	       }
	    if (eof == 1)
	       break;
	    n++;				/* set proper record number */
            line_no++;
	    if (!cdb_batchmode)
	       fb_gcounter(n);
	    fb_w_write(0, " ");
	    size++;
            if (cdb_secure){
               fb_nextwrite(0, perm);		      	/* permissions */
               size += 10;
               }
	    fb_w_write(0, "\000");			/* EOS = e-o-field */
	    fb_w_write(0, cdb_S_EOREC);			/* end of record */
	    size += 2;
	    lastc = NULL;

	    rlen = MAX(size, (long) (FB_MAPREC_SIZE + 2L));
	    if (size < rlen){			/* fill hole if needed */
	       while (++size < rlen)
	          fb_w_write(0, cdb_S_FILLC);	/* end of record */
	       fb_w_write(0, cdb_S_FILLE);	/* rec[rlen] = fille */
	       }

	    w_avail = avail;
	    w_rlen = rlen;
            if (cdb_dbase_byteorder != cdb_cpu_byteorder){
               M_32_SWAP(w_avail);
               M_32_SWAP(w_rlen);
               }
	    /* fb_put the map entry */
	    if (fb_w_writen(1, (char *) &w_avail, FB_SLONG) != 1){
	       fb_serror(FB_WRITE_ERROR, SYSMSG[S_BAD_MAP], NIL);
	       return(FB_ERROR);
	       }
	    if (fb_w_writen(1, (char *) &w_rlen, FB_SLONG) != 1){
	       fb_serror(FB_WRITE_ERROR, SYSMSG[S_BAD_MAP], NIL);
	       return(FB_ERROR);
	       }
	    avail += rlen;
	    }

         fb_wflush(2);
	 /* patch up fb_database header */
	 if (fb_puthead(cdb_db->fd, n, cdb_db->delcnt) == FB_ERROR)
	    fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_HEADER], cdb_db->dbase);

	 /* patch up fb_database header */
	 if (fb_putmap(cdb_db->mfd, 0L, avail, 0L, 0L)==FB_ERROR){
	    fb_serror(FB_IO_ERROR, SYSMSG[S_BAD_DATA], SYSMSG[S_BAD_MAP]);
	    return(FB_ERROR);
	    }
         return(FB_AOK);
      }

/* 
 *  simplefield - get a single simple fb_field from file fd (comma sep format)
 *     return value in cdb_bfld.
 */
 
   static simplefield()
      
         {
	    char *p, c;
	    int qlev, eof;  	/* eof is end of fb_field */
	       
	    qlev = 0; 
	    eof = 0;
	    p = cdb_bfld;
	    for (;;){			/* use cdb_bfld to gather */
	       if (fb_nextread(&c) == 0)
	          return(EOF);
	       if (c != '\n' && c != '\t' && c != ' ' && !isprint(c))
		  c = CHAR_STAR;
	       if (c == CHAR_BACKSLASH){	/* covers \\ and \" .. */
		  fb_nextread(&c);
		  if (c == 'n')
		     c = CHAR_NEWLINE;
		  }
	       else if (c == separator){
		  if (qlev == 0 || qlev == 2)  
		     eof = 1;		/* must be end of fb_field */
		  }
	       else if (c == CHAR_NEWLINE){
		  eof = 1;
		  }
	       else if (c == CHAR_QUOTE){
		  qlev++;
		  continue;			/* jump out */
		  }
	       if (eof)
	          break;
	       *p++ = c;			/* store character */
	       size++;
	       }
	    *p = NULL;				/* return field in cdb_bfld */
	    lastc = c;
	    return(FB_AOK);
	 }

/* 
 *  finit - intiialize --- merely check for existance. ask about overwrite 
 */
 
   static finit(argc, argv)
      int argc;
      char *argv[];
      
      {
         char str[FB_MAXLINE], fname[FB_MAXNAME];
	 int i;
	 
	 if (!append)
            if (access(cdb_db->dbase, 0) != FB_ERROR)
               fb_xerror(FB_MESSAGE, "Database already exists: ",
                  cdb_db->dbase);
	 fname[0] = NULL;
	 for (i = 1; i < argc; i++)
	    if (argv[i][0] == '-'){
	       if (argv[i][1] == 'd' || argv[i][1] == 'i' ||
		   argv[i][1] == 's' || argv[i][1] == 'c')
	          i++;
	       }
	    else{
	       strcpy(fname, argv[i]);
	       break;
	       }
	 if (fname[0] == NULL){		/* requires full path name */
	    fb_basename(fname, cdb_db->dbase);	/* calc default file name */
	    sprintf(str, "%s%s", fname, EMITEXTENSION);
	    if (fb_getfilename(fname, "Load File Name: ", str) == FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
	 sprintf(str, "from %s", fname);
	 fb_scrstat(str);
	 fb_r_init(fb_mustopen(fname, READ));
      }

/* 
 *  usage message 
 */
 
   static void usage()
      {
         fb_xerror(FB_MESSAGE, 
	    "usage: dbload [-b] [-y] [-a] [-l] [-c sep] [-d dbase] [LOADFILE]",
             NIL);
      }
